public/app/plugins/datasource/loki/query_utils.ts

```diff
@@ -15,61 +16,54 @@ export function formatQuery(selector: string | undefined): string {
  * E.g., `{} |= foo |=bar != baz` returns `['foo', 'bar']`.
  */
 export function getHighlighterExpressionsFromQuery(input: string): string[] {
-  let expression = input;
   const results = [];
 
-  // Consume filter expression from left to right
-  while (expression) {
-    const filterStart = expression.search(/\|=|\|~|!=|!~/);
-    // Nothing more to search
-    if (filterStart === -1) {
-      break;
-    }
-    // Drop terms for negative filters
-    const filterOperator = expression.slice(filterStart, filterStart + 2);
-    const skip = expression.slice(filterStart).search(/!=|!~/) === 0;
-    expression = expression.slice(filterStart + 2);
-    if (skip) {
+  const tree = parser.parse(input);
+  const filters: SyntaxNode[] = [];
+  tree.iterate({
+    enter: (type, from, to, get): void => {
+      if (type.id === LineFilter) {
+        filters.push(get());
+      }
+    },
+  });
+
+  for (let filter of filters) {
+    const pipeExact = filter.getChild(Filter)?.getChild(PipeExact);
+    const pipeMatch = filter.getChild(Filter)?.getChild(PipeMatch);
+    const string = filter.getChild(String);
+
+    if ((!pipeExact && !pipeMatch) || !string) {
       continue;
     }
-    // Check if there is more chained, by just looking for the next pipe-operator
-    const filterEnd = expression.search(/\|/);
-    let filterTerm;
-    if (filterEnd === -1) {
-      filterTerm = expression.trim();
-    } else {
-      filterTerm = expression.slice(0, filterEnd).trim();
-      expression = expression.slice(filterEnd);
-    }
+
+    const filterTerm = input.substring(string.from, string.to);
 
     const quotedTerm = filterTerm.match(/"(.*?)"/);
     const backtickedTerm = filterTerm.match(/`(.*?)`/);
     const term = quotedTerm || backtickedTerm;
```

Comment:
I noticed, that this will break the highlighting if we are filtering for a string containing quotes inside backticks:
![image](https://user-images.githubusercontent.com/8092184/183392023-891496da-b63e-4d04-b479-db16b26f1d3c.png)

I guess we can remove the regex matching here, determine the `filterTerm` by using `input.substring(string.from+1, string.to-1);` and `backtickedTerm` by just checking `input.substring(string.from, string.from+1);`?

Feels a bit hacky but should do the thing. Or what do you think @ivanahuckova and @matyax?

Reply:
Yeah, nice edge case catch. We should add test for this scenario. 

So the issue here is that in this case we match both `backtickedTerm` and  `quotedTerm` and because `term = quotedTerm || backtickedTerm`, it is going to match the `quotedTerm` inside. 

I am wondering if we should just determine if it is  `backtickedTerm` or `quotedTerm` based on `filterTerm[0]` character? It is always going to be just ` or ".

Reply:
Yea, the regex matching there will always cause some kind of issues, because quoted terms can contain backticks, and backticked terms can contain quotes.

> I am wondering if we should just determine if it is `backtickedTerm` or `quotedTerm` based on `filterTerm[0]` character? It is always going to be just ` or ".

Yea, that would essentially be the same as I was trying to say with 
> `backtickedTerm` by just checking `input.substring(string.from, string.from+1);`?

Reply:
Good catch! If I read the code correctly, the previous implementation also suffered from this edgecase, since that part was unchanged?

Anyway, in the last commit https://github.com/grafana/grafana/pull/53316/commits/977ab6ee77b7e258a5a223e7451065481564f0b3 I propose an alternative approach, where I updated the regex with beginning and end delimiters, to catch the actual quote type used for the whole term. Let me know what you think, and thanks for the feedback!

Reply:
Yep, definitely the previous implementation suffered from this as well.

Hm, https://github.com/grafana/grafana/commit/977ab6ee77b7e258a5a223e7451065481564f0b3 is fixing this, yea. I am unsure if I really prefer the RegEx matching over character matching just to determine, if it is a backticked filterterm. Performance-wise (even though it is not a super big issue) I would prefer checking based on `filterTerm[0]`, but we could also just stick to the solution now.

Reply:
We would also need to check `filterTerm[filterTerm.length - 1]` to be sure the term is properly quoted, right? Otherwise `"term" and `\`term` would incorrectly match.

I think the regex abstracts all those conditionals in a more readable way, opposite to the string-comparison approach.

Regarding performance, do you foresee any particular issue with it? Is this a function that gets called often?

Reply:
> We would also need to check filterTerm[filterTerm.length - 1] to be sure the term is properly quoted, right?

If it wasn't properly quoted, then parser shouldn't return `String` as string should be sterted-ended with the same set of quotes. But it does, so I am looking at the parser, trying to fix it these.

Reply:
Perfect, thanks!

Reply:
Updated.

Reply:
Btw here is the fix https://github.com/grafana/lezer-logql/pull/14. When it is merged, it will automatically add this new updated version `0.0.16` to `npm` and then we can update it in this `grafana/grafana` repo.

Reply:
I think with the parser-change we don't need to check the last character and can just go with the first character as backtick-or-not determination, right?

Reply:
Version `0.0.16` is out now https://www.npmjs.com/package/@grafana/lezer-logql/v/0.0.16.

Reply:
By the way, I'm waiting for your re-approval after the recent changes to merge. Not sure why it did not dismiss the original approve with further commits.